lib/cext/ruby/ruby.h

```diff
@@ -2215,6 +2118,319 @@ void ruby_sig_finalize(void);
+      optional = *formatp - '0';
+      formatp++;
+    }
+  }
+
+  if (*formatp == '*') {
+    rest = true;
+    formatp++;
+  } else {
+    rest = false;
+  }
+
+  if (isdigit(*formatp)) {
+    post = *formatp - '0';
+    formatp++;
+  }
+
+  if (*formatp == ':') {
+    kwargs = true;
+    formatp++;
+  } else {
+    kwargs = false;
+  }
+
+  if (*formatp == '&') {
+    block = true;
+    formatp++;
+  } else {
+    block = false;
+  }
+
+  if (*formatp != ' ') {
+    rb_raise(rb_eArgError, "bad rb_scan_args format");
+  }
+
+  // Check we have enough arguments
+
+  if (pre + post > argc) {
+    rb_raise(rb_eArgError, "not enough arguments for required");
+  }
+
+  // Read arguments
+
+  int argn = 0;
+  int valuen = 1; // We've numbered the v parameters from 1
+  bool taken_rest = false;
+  bool taken_block = false;
+  bool taken_kwargs = false;
+  bool erased_kwargs = false;
+  bool found_kwargs = false;
+
+  if (rest && kwargs && !truffle_invoke_b(RUBY_CEXT, "test_kwargs", argv[argc - 1], Qfalse)) {
+    kwargs = false;
+    erased_kwargs = true;
+  }
+
+  int trailing = post;
+
+  if (kwargs) {
+    trailing++;
+  }
+
+  while (true) {
+    // Get the next argument
+
+    VALUE arg;
+
+    if (pre > 0 || optional > 0) {
+      if (argn < argc - trailing) {
+        arg = argv[argn];
+        argn++;
+      } else {
+        arg = Qnil;
+      }
+
+      if (pre > 0) {
+        pre--;
+      } else {
+        optional--;
+      }
+    } else if (rest && !taken_rest) {
+      arg = rb_ary_new();
+      while (argn < argc - trailing) {
+        rb_ary_push(arg, argv[argn]);
+        argn++;
+      }
+      taken_rest = true;
+    } else if (post > 0) {
+      arg = argv[argn];
+      argn++;
+      post--;
+    } else if (kwargs && !taken_kwargs) {
+       if (argn < argc) {
+        arg = argv[argn];
+        truffle_invoke(RUBY_CEXT, "test_kwargs", argv[argn], Qtrue);
+        argn++;
+        found_kwargs = true;
+      } else {
+        arg = Qnil;
+      }
+      taken_kwargs = true;
+    } else if (erased_kwargs && !taken_kwargs) {
+      arg = Qnil;
+      taken_kwargs = true;
+    } else if (block && !taken_block) {
+      if (rb_block_given_p()) {
+        arg = rb_block_proc();
+      } else {
+        arg = Qnil;
+      }
+      taken_block = true;
+    } else {
+      break;
+    }
+
+    // Put the argument into the current value pointer
+
+    // Don't assign the correct v to a temporary VALUE* and then assign arg to it - this doesn't optimise well
+
+    switch (valuen) {
+      case 1: *v1 = arg; break;
+      case 2: *v2 = arg; break;
+      case 3: *v3 = arg; break;
+      case 4: *v4 = arg; break;
+      case 5: *v5 = arg; break;
+      case 6: *v6 = arg; break;
+      case 7: *v7 = arg; break;
+      case 8: *v8 = arg; break;
+      case 9: *v9 = arg; break;
+      case 10: *v10 = arg; break;
+    }
+
+    valuen++;
+  }
+
+  if (found_kwargs) {
+    return argc - 1;
+  } else {
+    return argc;
+  }
+}
+
+
+
```

Comment:
I would split the huge number of additions we have here in another file like `lib/cext/truffle/ruby.h`, so then the diff is just one `#include` line.